grin/test/Reducer/Interpreter/Definitional/CibSpec.hs

```diff
@@ -0,0 +1,115 @@
+module Reducer.Interpreter.Definitional.CibSpec where
+
+import Data.Monoid
+import Control.Exception
+import Test.Hspec
+import Reducer.Interpreter.Base
+import Reducer.Interpreter.Definitional.Cib
+import Reducer.Interpreter.Definitional.Internal
+import Data.Functor.Foldable
+import Grin.ExtendedSyntax.Syntax
+
+import qualified Reducer.Interpreter.Store as Store
+import qualified Data.Map as Map
+
+
+test :: IO ()
+test = hspec spec
+
+spec :: Spec
+spec = describe "Counting Immutable Beans operands" $ do
+  it "Increment reference counters" $ do
+    let exp =
```

Comment:
Would it be possible to parse grin-like code into this ADT using template-haskell? It would make the tests much more readable.

Reply:
It is possible, but we shouldn't write operations from CIB paper be hand. This is a stepping stone for the implementation of CIB paper. I killed two birds with one stone using the CIB instrumentation to implement the extensible abstract GRIN interpreter framework.

Reply:
I see, thanks.